File: 69-sqrtx/sqrtx.py
class Solution:
    def mySqrt(self, x: int) -> int:
        # Handle small x directly
        if x < 2:
            return x

        left, right = 1, x // 2  # sqrt(x) <= x//2 for x >= 2
        ans = 1  # will track floor sqrt

        while left <= right:
            mid = (left + right) // 2

            # Compare mid * mid with x without overflow
            if mid <= x // mid:
                ans = mid           # mid^2 <= x, mid is a candidate
                left = mid + 1      # try larger
            else:
                right = mid - 1     # mid^2 > x, try smaller

        return ans

File: 69-sqrtx/sqrtx.java
public class Solution {
    public int mySqrt(int x) {
        if (x < 2) return x;

        int left = 1, right = x / 2 + 1; // sqrt(x) <= x/2+1 for x>=2
        int ans = 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (mid <= x / mid) {   // mid*mid <= x without overflow
                ans = mid;          // mid is feasible
                left = mid + 1;     // try bigger
            } else {
                right = mid - 1;    // too big
            }
        }
        return ans;
    }
}
